resource.go

```diff
@@ -60,13 +82,79 @@ func getResourceField(v interface{}) (f *structs.Field, err error) {
 	return
 }
 
+func getIdFromMap(m map[string]interface{}) (ret int, err error) {
+	if val, ok := m["id"].(int); !ok {
+		ret = 0
+	} else {
+		ret = val
+	}
+	return
+}
+
 // Retrieve Id field from interface.
 func resourceId(v interface{}) (ret int, err error) {
 	var f *structs.Field
+	invalidTypeErr := errors.New("Not a struct, map, or pointer to either")
+	ty := reflect.TypeOf(v)
 
-	if f, err = getResourceField(v); err == nil {
-		ret = f.Value().(int)
+	switch ty.Kind() {
+	case reflect.Map:
+		if m, ok := v.(map[string]interface{}); !ok {
+			err = errors.New("Invalid map")
+		} else {
+			return getIdFromMap(m)
+		}
+	case reflect.Struct:
+		if f, err = getResourceField(v); err == nil {
+			ret = f.Value().(int)
+		}
+	case reflect.Ptr:
+		return resourceId(reflect.Indirect(reflect.ValueOf(v)).Interface())
+	default:
+		err = invalidTypeErr
 	}
 
 	return
 }
+
+// Determine whether the struct's id will be omitted in json encoding.
+func isIdOmitted(v interface{}) (bool, error) {
+
+	if structs.IsStruct(v) {
+		if f, err := getResourceField(v); err != nil {
+			return false, err
+		} else if jsonTag := f.Tag("json"); jsonTag == "" {
+			// No json tag means the id field won't be ommitted.
+			return false, nil
+		} else {
+			// getResourceField() ensures this is an int.
+			id := f.Value().(int)
+			// Json tags are comma separated. 'omitempty' means id == 0 -> id field
+			// not included in generated json. Also note spacing, like "id,
+			// omitempty" is significant and prevents a tag from taking effect so no
+			// need for trimming.
+			//
+			// See http://golang.org/pkg/encoding/json/#Marshal
+			for _, field := range strings.Split(jsonTag, ",") {
```

Comment:
This is indeed part of your original code, haven't touched it, only moved it here as I perceived it was directly related to the "resource" - I can move it back though